dbms/src/Columns/ColumnVector.cpp

```diff
@@ -203,6 +203,69 @@ void ColumnVector<T>::insertRangeFrom(const IColumn & src, size_t start, size_t
     memcpy(&data[old_size], &src_vec.data[start], length * sizeof(data[0]));
 }
 
+namespace
+{
+
+/// If mask is a number of this kind: [0]*[1]* function returns the length of the cluster of 1s.
+/// Otherwise it returns the special value: 0xFF.
+UInt8 prefixToCopy(UInt64 mask)
+{
+    if (mask == 0)
+        return 0;
+    if (mask == static_cast<UInt64>(-1))
+        return 64;
+    /// Row with index 0 correspond to the least significant bit.
+    /// So the length of the prefix to copy is 64 - #(leading zeroes).
+    const UInt64 leading_zeroes = __builtin_clzll(mask);
+    if (mask == ((static_cast<UInt64>(-1) << leading_zeroes) >> leading_zeroes))
+        return 64 - leading_zeroes;
+    else
+        return 0xFF;
+}
+
+uint8_t suffixToCopy(UInt64 mask)
+{
+    const auto prefix_to_copy = prefixToCopy(~mask);
+    return prefix_to_copy >= 64 ? prefix_to_copy : 64 - prefix_to_copy;
+}
+
+} // namespace
+
+template <typename T, typename Container, size_t SIMD_BYTES>
+inline void doFilterAligned(const UInt8 *& filt_pos, const UInt8 *& filt_end_aligned, const T *& data_pos, Container & res_data)
+{
+    while (filt_pos < filt_end_aligned)
+    {
+        UInt64 mask = ToBits64(filt_pos);
+        const UInt8 prefix_to_copy = prefixToCopy(mask);
+
+        if (0xFF != prefix_to_copy)
+        {
+            res_data.insert(data_pos, data_pos + prefix_to_copy);
```

Comment:
Is it possible to find every pair of start / end position of continuous section to copy?

Reply:
Seems do not have such function to achieve this.

Reply:
``` cpp
// 1
UInt64 mask = ToBits64(filt_pos);
const UInt8 prefix_to_copy = prefixToCopy(mask);
if (0xFF != prefix_to_copy)
{
    res_data.insert(data_pos, data_pos + prefix_to_copy);
}
else
{
    const uint8_t suffix_to_copy = suffixToCopy(mask);
    if (0xFF != suffix_to_copy)
    {
        res_data.insert(data_pos + SIMD_BYTES - suffix_to_copy, data_pos + SIMD_BYTES);
    }
    else
    {
        while (mask)
        {
            size_t index = __builtin_ctzll(mask);
            res_data.push_back(data_pos[index]);
            mask &= mask - 1;
        }
    }
}

// 2
UInt64 mask = ToBits64(filt_pos);
if (0xFFFFFFFFFFFFFFFFULL == mask)
{
    res_data.insert(data_pos, data_pos + SIMD_BYTES);
}
else
{
    int processed_bytes = 0;
    while (mask)
    {
        const int prefix_to_copy = __builtin_ctzll(~mask);
        if (prefix_to_copy)
        {
            res_data.insert(data_pos + processed_bytes, data_pos + processed_bytes + prefix_to_copy);
            processed_bytes += prefix_to_copy;
            mask >>= prefix_to_copy;
        }
        else
        {
            const int suffix_to_pass = __builtin_ctzll(mask);
            processed_bytes += suffix_to_pass;
            mask >>= suffix_to_pass;
        }
    }
}
```

```shell 
# 1
[qiuyang@172-16-5-85 ~/projects/tiflash/cmake-build-release 13:25:44]$ ./dbms/src/Columns/tests/column_vector_perftest int64 filter 10000 1 30
Test int64-filter rows=10000 columns=1 seconds=30
Filter/s: 156726    
[qiuyang@172-16-5-85 ~/projects/tiflash/cmake-build-release 13:26:25]$ ./dbms/src/Columns/tests/column_vector_perftest int64 filter 10000 10 30
Test int64-filter rows=10000 columns=10 seconds=30
Filter/s: 336723    
[qiuyang@172-16-5-85 ~/projects/tiflash/cmake-build-release 13:27:06]$ ./dbms/src/Columns/tests/column_vector_perftest int64 filter 10000 50 30
Test int64-filter rows=10000 columns=50 seconds=30
Filter/s: 632002    
[qiuyang@172-16-5-85 ~/projects/tiflash/cmake-build-release 13:27:45]$ ./dbms/src/Columns/tests/column_vector_perftest int64 filter 10000 100 30
Test int64-filter rows=10000 columns=100 seconds=30
Filter/s: 624887    
[qiuyang@172-16-5-85 ~/projects/tiflash/cmake-build-release 13:28:29]$ ./dbms/src/Columns/tests/column_vector_perftest int64 filter 10000 101 30
Test int64-filter rows=10000 columns=101 seconds=30
Filter/s: 526091    

# 2
[qiuyang@172-16-5-85 ~/projects/tiflash/cmake-build-release 13:17:21]$ ninja column_vector_perftest
[0/2] Re-checking globbed directories...
[3/3] Linking CXX executable dbms/src/Columns/tests/column_vector_perftest
[qiuyang@172-16-5-85 ~/projects/tiflash/cmake-build-release 13:20:26]$ ./dbms/src/Columns/tests/column_vector_perftest int64 filter 10000 1 30
Test int64-filter rows=10000 columns=1 seconds=30
Filter/s: 151959    
[qiuyang@172-16-5-85 ~/projects/tiflash/cmake-build-release 13:20:58]$ ./dbms/src/Columns/tests/column_vector_perftest int64 filter 10000 10 30
Test int64-filter rows=10000 columns=10 seconds=30
Filter/s: 95072     
[qiuyang@172-16-5-85 ~/projects/tiflash/cmake-build-release 13:22:07]$ ./dbms/src/Columns/tests/column_vector_perftest int64 filter 10000 50 30
Test int64-filter rows=10000 columns=50 seconds=30
Filter/s: 328683    
[qiuyang@172-16-5-85 ~/projects/tiflash/cmake-build-release 13:22:41]$ ./dbms/src/Columns/tests/column_vector_perftest int64 filter 10000 100 30
Test int64-filter rows=10000 columns=100 seconds=30
Filter/s: 474129    
[qiuyang@172-16-5-85 ~/projects/tiflash/cmake-build-release 13:23:21]$ ./dbms/src/Columns/tests/column_vector_perftest int64 filter 10000 101 30
Test int64-filter rows=10000 columns=101 seconds=30
Filter/s: 1030307   
```